File: src/Graphics/swapchain.rs
use crate::Application::appcfg::*;
use crate::Application::applicationdata::ApplicationData;
use crate::Application::queuefamilyindices::QueueFamilyIndices;
use ::winit::window::{Window, WindowBuilder};
use anyhow::{anyhow, Ok, Result};
use vulkanalia::loader::{LibloadingLoader, LIBRARY};
use vulkanalia::prelude::v1_0::*;
use vulkanalia::vk::ExtDebugUtilsExtension;
use vulkanalia::vk::KhrSurfaceExtension;
use vulkanalia::vk::KhrSwapchainExtension;

#[derive(Clone, Debug)]
pub struct SwapchainSupport {
    pub capabilities: vk::SurfaceCapabilitiesKHR,
    pub formats: Vec<vk::SurfaceFormatKHR>,
    pub present_modes: Vec<vk::PresentModeKHR>,
}

impl SwapchainSupport {
    pub unsafe fn get(
        instance: &Instance,
        data: &ApplicationData,
        physical_device: vk::PhysicalDevice,
    ) -> Result<Self> {
        Ok(Self {
            capabilities: instance
                .get_physical_device_surface_capabilities_khr(physical_device, data.surface)?,
            formats: instance
                .get_physical_device_surface_formats_khr(physical_device, data.surface)?,
            present_modes: instance
                .get_physical_device_surface_present_modes_khr(physical_device, data.surface)?,
        })
    }
}

#[derive(Default, Clone, Debug)]
pub struct Swapchain {}

impl Swapchain {
    pub unsafe fn create_swapchain(
        window: &Window,
        instance: &Instance,
        device: &Device,
        data: &mut ApplicationData,
    ) -> Result<()> {
        let indices = QueueFamilyIndices::get(instance, data, data.physical_device)?;
        let support = SwapchainSupport::get(instance, data, data.physical_device)?;

        let surface_format = Self::get_swapchain_surface_format(&support.formats);
        let present_mode = Self::get_swapchain_present_mode(&support.present_modes);
        let extent = Self::get_swapchain_extent(window, support.capabilities);

        data.swapchain_format = surface_format.format;
        data.swapchain_extent = extent;

        let mut image_count = support.capabilities.min_image_count + 1;
        if support.capabilities.max_image_count != 0
            && image_count > support.capabilities.max_image_count
        {
            image_count = support.capabilities.max_image_count;
        }

        let mut queue_family_indices = vec![];
        let image_sharing_mode = if indices.graphics != indices.present {
            queue_family_indices.push(indices.graphics);
            queue_family_indices.push(indices.present);
            vk::SharingMode::CONCURRENT
        } else {
            vk::SharingMode::EXCLUSIVE
        };

        let info = vk::SwapchainCreateInfoKHR::builder()
            .surface(data.surface)
            .min_image_count(image_count)
            .image_format(surface_format.format)
            .image_color_space(surface_format.color_space)
            .image_extent(extent)
            .image_array_layers(1)
            .image_usage(vk::ImageUsageFlags::COLOR_ATTACHMENT)
            .image_sharing_mode(image_sharing_mode)
            .queue_family_indices(&queue_family_indices)
            .pre_transform(support.capabilities.current_transform)
            .composite_alpha(vk::CompositeAlphaFlagsKHR::OPAQUE)
            .present_mode(present_mode)
            .clipped(true)
            .old_swapchain(vk::SwapchainKHR::null());

        data.swapchain = device.create_swapchain_khr(&info, None)?;
        data.swapchain_images = device.get_swapchain_images_khr(data.swapchain)?;

        Ok(())
    }

    fn get_swapchain_surface_format(formats: &[vk::SurfaceFormatKHR]) -> vk::SurfaceFormatKHR {
        formats
            .iter()
            .cloned()
            .find(|f| {
                f.format == vk::Format::B8G8R8A8_SRGB
                    && f.color_space == vk::ColorSpaceKHR::SRGB_NONLINEAR
            })
            .unwrap_or_else(|| formats[0])
    }

    fn get_swapchain_present_mode(present_modes: &[vk::PresentModeKHR]) -> vk::PresentModeKHR {
        present_modes
            .iter()
            .cloned()
            .find(|m| *m == vk::PresentModeKHR::MAILBOX)
            .unwrap_or(vk::PresentModeKHR::FIFO)
    }

    fn get_swapchain_extent(
        window: &Window,
        capabilities: vk::SurfaceCapabilitiesKHR,
    ) -> vk::Extent2D {
        if capabilities.current_extent.width != u32::max_value() {
            capabilities.current_extent
        } else {
            let size = window.inner_size();
            let clamp = |min: u32, max: u32, v: u32| min.max(max.min(v));
            vk::Extent2D::builder()
                .width(clamp(
                    capabilities.min_image_extent.width,
                    capabilities.max_image_extent.width,
                    size.width,
                ))
                .height(clamp(
                    capabilities.min_image_extent.height,
                    capabilities.max_image_extent.height,
                    size.height,
                ))
                .build()
        }
    }

    pub unsafe fn create_swapchain_image_views(
        device: &Device,
        data: &mut ApplicationData,
    ) -> Result<()> {
        data.swapchain_image_views = data
            .swapchain_images
            .iter()
            .map(|i| {
                let components = vk::ComponentMapping::builder()
                    .r(vk::ComponentSwizzle::IDENTITY)
                    .g(vk::ComponentSwizzle::IDENTITY)
                    .b(vk::ComponentSwizzle::IDENTITY)
                    .a(vk::ComponentSwizzle::IDENTITY);

                let subresource_range = vk::ImageSubresourceRange::builder()
                    .aspect_mask(vk::ImageAspectFlags::COLOR)
                    .base_mip_level(0)
                    .level_count(1)
                    .base_array_layer(0)
                    .layer_count(1);

                let info = vk::ImageViewCreateInfo::builder()
                    .image(*i)
                    .view_type(vk::ImageViewType::_2D)
                    .format(data.swapchain_format)
                    .components(components)
                    .subresource_range(subresource_range);

                device.create_image_view(&info, None)
            })
            .collect::<Result<Vec<_>, _>>()?;

        Ok(())
    }

    pub unsafe fn create_sync_objects(device: &Device, data: &mut ApplicationData) -> Result<()> {
        let semaphore_info = vk::SemaphoreCreateInfo::builder();
        let fence_info = vk::FenceCreateInfo::builder().flags(vk::FenceCreateFlags::SIGNALED);

        for _ in 0..MAX_FRAMES_IN_FLIGHT {
            data.image_available_semaphores
                .push(device.create_semaphore(&semaphore_info, None)?);

            data.render_finished_semaphores
                .push(device.create_semaphore(&semaphore_info, None)?);

            data.in_flight_fences
                .push(device.create_fence(&fence_info, None)?);
        }

        data.images_in_flight = data
            .swapchain_images
            .iter()
            .map(|_| vk::Fence::null())
            .collect();

        Ok(())
    }
}

File: engine/src/vulkan/swapchain/swapchain.cpp
#include "swapchain.hpp"
#include <vulkan/vulkan.h>
#define GLFW_INCLUDE_VULKAN
#include <GLFW/glfw3.h>
#include <cstdint>
#include <limits>
#include <algorithm>
#include "../../core/instance/instance.hpp"
#include "../physicaldevice/physicaldevice.hpp"
#include "../../defines.hpp"

namespace gloria::vk {
	SwapChain::SwapChain() {
	}

	SwapChain::~SwapChain() {
	}

	void SwapChain::init() {
		createSwapChain();
		createImageViews();
	}

	void SwapChain::createSwapChain() {
		SwapChainSupportDetails swapChainSupport = core::Instance::get().getVkInstance().getPhysicalDevice().querySwapChainSupport(
			core::Instance::get().getVkInstance().getPhysicalDevice().get());

		VkSurfaceFormatKHR surfaceFormat = selectSwapSurfaceFormat(swapChainSupport.formats);
		VkPresentModeKHR presentMode = selectSwapPresentMode(swapChainSupport.presentModes);
		VkExtent2D extent = selectSwapExtent(swapChainSupport.capabilities);

		std::uint32_t imageCount = swapChainSupport.capabilities.minImageCount + mSwapChainImageAmount; // min amount + the amount we want

		if (swapChainSupport.capabilities.maxImageCount > 0 && imageCount > swapChainSupport.capabilities.maxImageCount) {
			imageCount = swapChainSupport.capabilities.maxImageCount;
		}

		VkSwapchainCreateInfoKHR createInfo = {
			.sType = VK_STRUCTURE_TYPE_SWAPCHAIN_CREATE_INFO_KHR,
			.surface = core::Instance::get().getVkInstance().getSurface().get(),
			.minImageCount = imageCount,
			.imageFormat = surfaceFormat.format,
			.imageColorSpace = surfaceFormat.colorSpace,
			.imageExtent = extent,
			.imageArrayLayers = 1,
			.imageUsage = VK_IMAGE_USAGE_COLOR_ATTACHMENT_BIT
		};

		swapChainImageFormat = surfaceFormat.format;
		swapChainExtent = extent;

		QueueFamilyIndices indices = findQueueFamilies(core::Instance::get().getVkInstance().getPhysicalDevice().get());
		std::uint32_t queueFamilyIndices[] = { indices.graphicsFamily.value(), indices.presentFamily.value() };

		if (indices.graphicsFamily != indices.presentFamily) {
			createInfo.imageSharingMode = VK_SHARING_MODE_CONCURRENT;
			createInfo.queueFamilyIndexCount = 2;
			createInfo.pQueueFamilyIndices = queueFamilyIndices;
		}
		else {
			createInfo.imageSharingMode = VK_SHARING_MODE_EXCLUSIVE;
			createInfo.queueFamilyIndexCount = 0;
			createInfo.pQueueFamilyIndices = nullptr;
		}

		createInfo.preTransform = swapChainSupport.capabilities.currentTransform;
		createInfo.compositeAlpha = VK_COMPOSITE_ALPHA_OPAQUE_BIT_KHR;
		createInfo.presentMode = presentMode;
		createInfo.clipped = VK_TRUE;
		createInfo.oldSwapchain = VK_NULL_HANDLE;

		VK_VALIDATE(vkCreateSwapchainKHR(core::Instance::get().getVkInstance().getLogicalDevice().get(), &createInfo, nullptr, &mSwapChain), "Failed to create swapchain");

		vkGetSwapchainImagesKHR(core::Instance::get().getVkInstance().getLogicalDevice().get(), mSwapChain, &imageCount, nullptr);
		swapChainImages.resize(imageCount);
		vkGetSwapchainImagesKHR(core::Instance::get().getVkInstance().getLogicalDevice().get(), mSwapChain, &imageCount, swapChainImages.data());

#ifdef DEBUG
		if (mSwapChain != VK_NULL_HANDLE)
			GL_CORE_INFO("Created swapchain");
#endif // DEBUG
	}

	void SwapChain::destroy() {
		for (auto imageView : swapChainImageViews) {
			vkDestroyImageView(core::Instance::get().getVkInstance().getLogicalDevice().get(), imageView, nullptr);
		}

		vkDestroySwapchainKHR(core::Instance::get().getVkInstance().getLogicalDevice().get(), mSwapChain, nullptr);
	}

	VkSwapchainKHR& SwapChain::get() {
		return mSwapChain;
	}

	void SwapChain::createImageViews() {
		swapChainImageViews.resize(swapChainImages.size());

		for (std::size_t i = 0; i < swapChainImages.size(); i++) {
			VkImageViewCreateInfo createInfo = {
				.sType = VK_STRUCTURE_TYPE_IMAGE_VIEW_CREATE_INFO,
				.image = swapChainImages[i],
				.viewType = VK_IMAGE_VIEW_TYPE_2D,
				.format = swapChainImageFormat,
				.components = {
					.r = VK_COMPONENT_SWIZZLE_IDENTITY,
					.g = VK_COMPONENT_SWIZZLE_IDENTITY,
					.b = VK_COMPONENT_SWIZZLE_IDENTITY,
					.a = VK_COMPONENT_SWIZZLE_IDENTITY
				},
				.subresourceRange = {
					.aspectMask = VK_IMAGE_ASPECT_COLOR_BIT,
					.baseMipLevel = 0,
					.levelCount = 1,
					.baseArrayLayer = 0,
					.layerCount = 1
				}
			};
			VK_VALIDATE(vkCreateImageView(core::Instance::get().getVkInstance().getLogicalDevice().get(), &createInfo, nullptr, &swapChainImageViews[i]), "Failed to create image views");

#ifdef DEBUG
			if (swapChainImageViews[i] != VK_NULL_HANDLE)
				GL_CORE_INFO("Created image view");
#endif // DEBUG
		}
	}

	void SwapChain::createSyncObjects(VkDevice device) {
		imageAvailableSemaphores.resize(MAX_FRAMES_IN_FLIGHT);
		renderFinishedSemaphores.resize(MAX_FRAMES_IN_FLIGHT);
		inFlightFences.resize(MAX_FRAMES_IN_FLIGHT);

		VkSemaphoreCreateInfo semaphoreInfo = {
	.sType = VK_STRUCTURE_TYPE_SEMAPHORE_CREATE_INFO
		};

		VkFenceCreateInfo fenceInfo = {
			.sType = VK_STRUCTURE_TYPE_FENCE_CREATE_INFO,
			.flags = VK_FENCE_CREATE_SIGNALED_BIT
		};

		for (std::size_t i = 0; i < MAX_FRAMES_IN_FLIGHT; i++) {
			VK_VALIDATE(vkCreateSemaphore(device, &semaphoreInfo, nullptr, &imageAvailableSemaphores[i]) ||
				vkCreateSemaphore(device, &semaphoreInfo, nullptr, &renderFinishedSemaphores[i]) ||
				vkCreateFence(device, &fenceInfo, nullptr, &inFlightFences[i]), "Failed to create semaphores");
		}
	}

	void SwapChain::recreateSwapchain() {
		int width = 0, height = 0;
		glfwGetFramebufferSize(core::Instance::get().getWindow().getRawWindow(), &width, &height);
		while (width == 0 || height == 0) {
			glfwGetFramebufferSize(core::Instance::get().getWindow().getRawWindow(), &width, &height);
			glfwWaitEvents();
		}

		vkDeviceWaitIdle(core::Instance::get().getVkInstance().getLogicalDevice().get());

		destroy();

		createSwapChain();
		createImageViews();
	}

	// check for our wanted surface format, if it cant be found we just return the best one.
	VkSurfaceFormatKHR SwapChain::selectSwapSurfaceFormat(const std::vector<VkSurfaceFormatKHR>& availableFormats) {
		for (const auto& availableFormat : availableFormats) {
			if (availableFormat.format == VK_FORMAT_B8G8R8A8_SRGB && availableFormat.colorSpace == VK_COLOR_SPACE_SRGB_NONLINEAR_KHR) {
				return availableFormat;
			}
		}
		return availableFormats[0];
	}

	// check for our wanted presentmode, if it cant be found we just return the standard one that always exists.
	VkPresentModeKHR SwapChain::selectSwapPresentMode(const std::vector< VkPresentModeKHR>& availablePresentModes) {
		for (const auto& availablePresentMode : availablePresentModes) {
			if (availablePresentMode == VK_PRESENT_MODE_MAILBOX_KHR) {
				return availablePresentMode;
			}
		}
		return VK_PRESENT_MODE_FIFO_KHR;
	}

	VkExtent2D SwapChain::selectSwapExtent(const VkSurfaceCapabilitiesKHR& capabilities) {
		if (capabilities.currentExtent.width != std::numeric_limits<std::uint32_t>::max()) {
			return capabilities.currentExtent;
		}
		else {
			int width, height;

			glfwGetFramebufferSize(core::Instance::get().getWindow().getRawWindow(), &width, &height);

			VkExtent2D correctedExtent = {
				static_cast<std::uint32_t>(width),
				static_cast<std::uint32_t>(height)
			};

			// clamp the values between the possible min, max
			correctedExtent.width = std::clamp(correctedExtent.width, capabilities.minImageExtent.width, capabilities.maxImageExtent.width);
			correctedExtent.height = std::clamp(correctedExtent.height, capabilities.minImageExtent.height, capabilities.maxImageExtent.height);

			return correctedExtent;
		}
	}
}